zilliqa/src/exec.rs

```diff
@@ -574,13 +598,27 @@ impl State {
                     success: false,
                     contract_address: None,
                     logs: Default::default(),
+                    traces: Default::default(),
                 })
             }
         }
     }
 
     // Apply the changes the EVM is requesting for
-    fn apply_delta(&mut self, applys: Vec<evm_ds::protos::evm_proto::Apply>) -> Result<()> {
+    fn apply_delta(
+        &mut self,
+        applys: Vec<evm_ds::protos::evm_proto::Apply>,
+        new_addresses: Vec<Address>,
+    ) -> Result<()> {
+        // these accounts have been 'created' by having funds sent to them. We need to create them with nonce 0
```

Comment:
we now require to know which addresses have been potentially created by a TX so that we can make sure that there is always an account that maps to the funding contract (so we can determine whether TXs 'from' are valid).

I actually would slightly like the idea of caching the balance of accounts too, that way you don't need to execute the evm for a balance check. what do people think?

Reply:
IMO we should avoid adding a cache before we've shown that executing the EVM to calculate balances is a bottleneck (and even then the answer might be to speed up our EVM).

Reply:
Regarding needing to know which accounts are 'created'. It strikes me as peculiar that we need to know. The mental model I have right now is `state: Map<Address, Account>` and accounts get accessed as `state.get(address).unwrap_or(Account::default_with_zero_balance_and_nonce())` (i.e. if an account doesn't exist in the map, we default it to the new account and if it gets mutated, it will be written back to state).

There are quite a few cases in which an account could have state, without native funds being sent to them (e.g. `CREATE2` deployments) and I'm apprehensive about us trying to enumerate those cases.

Would a generic approach in which we assume a default account if it doesn't exist work?

Reply:
The cache wouldn't be purely (or mainly) for performance reasons but for scenarios like:
- to test if you have enough gas to execute a tx you can do it without evm execution which reduces logs
- If you are debugging more intuitive/easy to dump the accounts + balances and inspect
- calls to eth_getBalance/gasPrice which are common you likely don't care about seeing the evm churning away for
-when the evm calls to ask what balances are it doesn't have to recurse by making another call

So it's just quality of life, but just an idea.

Wrt knowing which accounts are 'created' - it is less complicated than that. All it is doing is if a transfer is made to a location, if there is currently no account there, make one. Again, it is just quality of life but I am not tied to it